File: loopx/control_plane/coordination/todo_continuation.ts
/** Explicit cross-agent, cross-session continuation over the current Todo.
 * This is a local workflow note, not a history, search or ownership protocol.
 * Phase 1: same-machine, different registered agents.
 *
 * Stage A supports two handoff styles:
 * - Legacy: rationale + source_refs (manual, compact)
 * - Rich context: work_summary + structured fields (approaches_tried,
 *   next_steps, files_touched, key_decisions, open_questions)
 * Both produce a typed continuation note validated by the shared
 * continuation_note.ts predicate. The rich form lets a source agent dump
 * complete working context so the target agent does not need to re-explain
 * what was done, what failed, and what comes next.
 */
import {stat} from "node:fs/promises";
import {isAbsolute, resolve, relative} from "node:path";
import type {JsonObject} from "../effect_program.ts";
import type {AuthorityStore} from "./authority_store.ts";
import {canonicalAuthorityObject, canonicalAuthoritySha256, requireAuthorityStoreId} from "./authority_store_codec.ts";
import {indexCoordinationProjection, validateCoordinationTodoReadModel} from "./coordination_projection.ts";
import {executeCoordinationTodoUpdate} from "./todo_update.ts";
import {executeCoordinationTodoClaim} from "./todo_claim.ts";
import {
  CONTINUATION_NOTE_MARKER,
  buildContinuationNote,
  computeContinuationTodoFacts,
  validateContinuationNote,
  validateRawContext,
  type ContinuationNoteContext,
  type ContinuationNoteValidation,
  type ContinuationNoteApproachTried,
  type ContinuationNoteFileTouched,
  type ContinuationNoteDecision,
} from "./continuation_note.ts";

const accepted = new Set(["applied", "replayed", "recovered", "no_change"]);
const reject = (reason_code: string, reason: string): JsonObject =>
  ({ok: false, status: "rejected", reason_code, reason});

function bounded(value: unknown, name: string, max: number): string {
  if (typeof value !== "string" || !value.trim() || value.length > max) {
    throw new Error(`${name} must be non-empty text of at most ${max} characters`);
  }
  return value;
}

// Excludes claimed_by so handoff verification ignores the expected ownership transfer.
function stableFacts(todo: JsonObject): string {
  const copy = {...todo};
  for (const key of ["note", "updated_at", "last_actor_agent_id", "claimed_by"]) delete copy[key];
  return canonicalAuthoritySha256(copy);
}

async function availability(workspace: unknown, artifacts: unknown): Promise<JsonObject> {
  if (typeof workspace !== "string" || !isAbsolute(workspace)) throw new Error("workspace must be absolute");
  if (!Array.isArray(artifacts) || artifacts.length > 8) throw new Error("at most eight artifacts are supported");
  const inspect = async (path: string, directory: boolean) => {
    try { const info = await stat(path); return directory ? info.isDirectory() : info.isFile(); }
    catch (error) { if (["ENOENT", "ENOTDIR"].includes((error as NodeJS.ErrnoException).code ?? "")) return false; throw error; }
  };
  const workspaceAvailable = await inspect(workspace, true);
  const rows = [];
  for (const value of artifacts) {
    const ref = bounded(value, "artifact", 240);
    if (isAbsolute(ref) || relative(workspace, resolve(workspace, ref)).split(/[\\/]/u).includes("..")) {
      throw new Error("artifacts must be workspace-relative paths");
    }
    rows.push({ref, available: await inspect(resolve(workspace, ref), false)});
  }
  return {workspace_available: workspaceAvailable, artifacts: rows,
    ready: workspaceAvailable && rows.every(row => row.available)};
}

function buildContextFromInput(input: JsonObject): ContinuationNoteContext {
  // Two supported shapes:
  //   1. CLI path: context wrapped in input.context (dedicated field).
  //      Only context-only keys are allowed — operational keys (action,
  //      agent_id, goal_id, etc.) are REJECTED, not ignored. This is the
  //      security boundary: a context file cannot override command,
  //      identity, or authority metadata.
  //   2. Direct TypeScript callers: context fields passed flat (backward
  //      compat with unit tests). Request keys are allowed but only
  //      context keys are extracted.
  const contextKeys = new Set(["work_summary", "rationale", "source_refs", "approaches_tried",
    "next_steps", "files_touched", "key_decisions", "open_questions"]);
  const requestKeys = new Set(["goal_id", "todo_id", "agent_id", "registered_agents", "action",
    "session_id", "operation_id", "expected_provider_revision", "workspace", "artifacts",
    "target_agent_id", "handoff_format", "runtime_root"]);
  // Resolve the raw context source.
  let rawContext: JsonObject;
  if (input.context !== undefined) {
    // CLI path: context is in the dedicated field. Validate it is an object.
    if (typeof input.context !== "object" || input.context === null || Array.isArray(input.context)) {
      throw new Error("context must be an object with context-only fields (work_summary, rationale, etc.)");
    }
    rawContext = input.context as JsonObject;
    // Only context keys are allowed. Reject ALL operational keys.
    for (const key of Object.keys(rawContext)) {
      if (!contextKeys.has(key)) {
        throw new Error(`unknown context field: ${key}. Context may only contain: ${[...contextKeys].join(", ")}`);
      }
    }
  } else {
    // Backward compat: flat fields. Allow both context and request keys.
    rawContext = input;
    for (const key of Object.keys(rawContext)) {
      if (!contextKeys.has(key) && !requestKeys.has(key)) {
        throw new Error(`unknown context field: ${key}`);
      }
    }
  }
  // Extract only the context fields.
  const context: JsonObject = {};
  for (const key of contextKeys) {
    if (key in rawContext) context[key] = rawContext[key];
  }
  // Validate the extracted context against the closed bounded schema. This
  // throws on the first violation — wrong type, bounds overflow, unknown
  // nested key, or missing required field — so the producer can never
  // silently sanitize bad input into a smaller but legal note.
  validateRawContext(context);
  // All fields are already validated — pass them through directly without
  // any second sanitization. The producer's output shape is guaranteed to
  // match what validateContinuationNote() will see later.
  return {
    work_summary: context.work_summary as string | undefined,
    rationale: context.rationale as string | undefined,
    source_refs: context.source_refs as readonly string[] | undefined,
    approaches_tried: context.approaches_tried as readonly ContinuationNoteApproachTried[] | undefined,
    next_steps: context.next_steps as readonly string[] | undefined,
    files_touched: context.files_touched as readonly ContinuationNoteFileTouched[] | undefined,
    key_decisions: context.key_decisions as readonly ContinuationNoteDecision[] | undefined,
    open_questions: context.open_questions as readonly string[] | undefined,
  };
}

export async function executeTodoContinuation(store: AuthorityStore, value: unknown): Promise<JsonObject> {
  const input = canonicalAuthorityObject(value, "continuation request");
  const goalId = requireAuthorityStoreId(input.goal_id, "goal id");
  const todoId = requireAuthorityStoreId(input.todo_id, "Todo id");
  const agentId = requireAuthorityStoreId(input.agent_id, "agent id");
  const sessionId = bounded(input.session_id, "session id", 160);
  if (!["prepare", "inspect", "adopt"].includes(String(input.action))) throw new Error("invalid continuation action");
  if (!Array.isArray(input.registered_agents) || !input.registered_agents.includes(agentId)) {
    return reject("actor_not_registered", "Use the registered lane bound to this host session");
  }
  const registered = input.registered_agents.map(v => requireAuthorityStoreId(v, "registered agent"));
  const targetAgentId = input.target_agent_id != null
    ? requireAuthorityStoreId(input.target_agent_id, "target agent id") : null;
  if (targetAgentId != null && !registered.includes(targetAgentId)) {
    return reject("target_agent_not_registered", "Target agent must be registered for this goal");
  }
  const head = await store.loadAuthority();
  if (head.status !== "loaded") return {ok: false, ...head};
  validateCoordinationTodoReadModel(head.head, goalId);
  const projection = indexCoordinationProjection(head.head, goalId);
  const todo = projection.todos.get(todoId);
  if (!todo) return reject("todo_not_found", "The stable Todo ID no longer exists; do not recreate it from text");
  if (todo.status !== "open" || todo.archive_state !== "active") return reject("todo_not_open", "Todo is no longer open and active");
  // The existing metadata writer cannot yet prove a lease-bearing note update.
  // Preserve its boundary instead of inventing a second lease/transfer protocol.
  if (![undefined, "legacy", "soft_claim"].includes(head.head.handoff_mode as string | undefined) || projection.leases.has(todoId)) {
    return reject("continuation_lease_unsupported", "Stage A supports lease-free local Todos only; use existing lease/handoff commands for leased work");
  }
  const common = {goal_id: goalId, todo_id: todoId, expected_role: "agent",
    actor_agent_id: agentId, registered_agents: registered, dry_run: false, now: new Date()};
  if (input.action === "prepare") {
    if (todo.claimed_by !== agentId) return reject("continuation_owner_required", "Only the current Todo owner can prepare a continuation note");
    const context = buildContextFromInput(input);
    const todoFacts = computeContinuationTodoFacts(todo);
    const noteObj = buildContinuationNote(context, sessionId, todoFacts);
    const note = JSON.stringify(noteObj);
    const result = await executeCoordinationTodoUpdate(store, {...common,
      operation_id: requireAuthorityStoreId(input.operation_id, "operation id"),
      expected_provider_revision: requireAuthorityStoreId(input.expected_provider_revision, "expected revision"),
      patch: {note}, clear_fields: []});
    const current = await store.loadAuthority();
    const verified = current.status === "loaded" &&
      indexCoordinationProjection(current.head, goalId).todos.get(todoId)?.note === note;
    return {ok: accepted.has(String(result.status)) && verified, action: "prepare", result,
      note_readback_verified: verified, goal_id: goalId, todo_id: todoId,
      provider_revision: current.status === "loaded" ? current.provider_revision : null};
  }
  const environment = await availability(input.workspace, input.artifacts ?? []);
  const noteValidation: ContinuationNoteValidation = validateContinuationNote(todo.note, computeContinuationTodoFacts(todo));
  const validNote = noteValidation.valid;
  const note = noteValidation.note;
  // Build a digest-friendly summary. The CLI can render this as a readable
  // handoff context for the target agent.
  const digest = validNote ? {
    work_summary: typeof note!.work_summary === "string" ? note!.work_summary : undefined,
    rationale: typeof note!.rationale === "string" ? note!.rationale : undefined,
    approaches_tried: Array.isArray(note!.approaches_tried) ? note!.approaches_tried : undefined,
    next_steps: Array.isArray(note!.next_steps) ? note!.next_steps : undefined,
    files_touched: Array.isArray(note!.files_touched) ? note!.files_touched : undefined,
    key_decisions: Array.isArray(note!.key_decisions) ? note!.key_decisions : undefined,
    open_questions: Array.isArray(note!.open_questions) ? note!.open_questions : undefined,
    source_refs: Array.isArray(note!.source_refs) ? note!.source_refs : undefined,
  } : undefined;
  const packet: JsonObject = {ok: true, action: "inspect", goal_id: goalId, todo_id: todoId,
    provider_revision: head.provider_revision, todo_status: todo.status, claimed_by: todo.claimed_by,
    note_state: validNote ? "current" : note?.kind === CONTINUATION_NOTE_MARKER ? "stale" : "missing",
    availability: environment, decision_rationale: validNote ? note!.rationale : null,
    evidence_refs: [`todo:${todoId}`, `revision:${head.provider_revision}`, ...(validNote && Array.isArray(note!.source_refs) ? note!.source_refs as string[] : [])],
    next_step: validNote && environment.ready === true ? todo.text : "Restore workspace/artifacts and ask the source session to prepare a current decision note",
    can_adopt: validNote && environment.ready === true && note!.source_session !== sessionId,
    digest};
  if (input.action === "inspect") return packet;
  if (!packet.can_adopt) return {...packet, ...reject("continuation_not_ready", "Inspect and repair the reported note/session/availability gap before adopting")};
  const adoptOwnerId = targetAgentId ?? agentId;
  const preClaimFacts = stableFacts(todo);
  // Construct a handoff transfer grant binding source owner, target owner,
  // todo id, exact revision, and current continuation-note fingerprint. The
  // note fingerprint is computed by the shared validateContinuationNote so the
  // final claim authority reuses the exact same canonicalization. An invalid
  // note yields empty noteFacts, which the final decision will reject.
  const transferGrant = adoptOwnerId !== todo.claimed_by ? {
    schema_version: "todo_transfer_grant_v0" as const,
    source_agent_id: todo.claimed_by as string,
    target_agent_id: adoptOwnerId,
    todo_id: todoId,
    expected_revision: requireAuthorityStoreId(input.expected_provider_revision, "expected revision"),
    continuation_note_facts: noteValidation.noteFacts,
  } : undefined;
  const result = await executeCoordinationTodoClaim(store, {...common, claimed_by: adoptOwnerId,
    operation_id: requireAuthorityStoreId(input.operation_id, "operation id"),
    expected_provider_revision: requireAuthorityStoreId(input.expected_provider_revision, "expected revision"),
    transfer_grant: transferGrant});
  // Historical receipt replay alone never proves current authority.
  const current = await store.loadAuthority();
  const currentTodo = current.status === "loaded" ? indexCoordinationProjection(current.head, goalId).todos.get(todoId) : undefined;
  const verified = currentTodo?.status === "open" && currentTodo.archive_state === "active" &&
    currentTodo.claimed_by === adoptOwnerId && currentTodo.note === todo.note && stableFacts(currentTodo) === preClaimFacts &&
    current.status === "loaded" && current.provider_revision === result.provider_revision;
  return {...packet, action: "adopt", ok: accepted.has(String(result.status)) && verified,
    claim: result, current_authority_verified: verified, target_agent_id: adoptOwnerId,
    provider_revision: current.status === "loaded" ? current.provider_revision : null};
}

File: loopx/cli_commands/todo_continuation.py
"""Host CLI adapter; all continuation and ownership decisions stay in TypeScript."""
from pathlib import Path
import json

from ..agent_registry import registered_agent_ids_from_registry
from ..history import load_registry
from ..paths import resolve_runtime_root
from ..control_plane.effect_runtime import effect_runtime_result

# Context-only keys: the --from-context JSON payload may ONLY contain
# these fields. Any operational key (action, agent_id, goal_id, etc.) is
# rejected before the payload reaches the TypeScript layer. This prevents
# a context file from overriding CLI-derived command/identity/authority
# metadata.
_CONTEXT_KEYS = {
    "work_summary", "rationale", "source_refs",
    "approaches_tried", "next_steps", "files_touched",
    "key_decisions", "open_questions",
}


def _validate_context_payload(context: object) -> dict:
    """Validate that a --from-context JSON value is an object with only
    context-only keys. Returns the validated dict. Raises ValueError with
    an actionable message on any violation."""
    if not isinstance(context, dict):
        raise ValueError(
            "--from-context must be a JSON object (work_summary, rationale, etc.); "
            f"got {type(context).__name__}"
        )
    unknown = sorted(set(context.keys()) - _CONTEXT_KEYS)
    if unknown:
        raise ValueError(
            f"unknown context field: {', '.join(unknown)}. "
            f"Allowed: {', '.join(sorted(_CONTEXT_KEYS))}"
        )
    return context


def _render_digest(payload: dict) -> str:
    """Render an inspect payload as a readable handoff digest for the target agent."""
    if not payload.get("ok"):
        return json.dumps(payload, indent=2, ensure_ascii=False)
    lines: list[str] = []
    lines.append("# Handoff Context")
    source = payload.get("claimed_by", "unknown")
    note_state = payload.get("note_state", "missing")
    lines.append(f"Source: {source} | Status: {note_state}")
    lines.append("")
    digest = payload.get("digest")
    if digest:
        if digest.get("work_summary"):
            lines.append("## Summary")
            lines.append(str(digest["work_summary"]))
            lines.append("")
        if digest.get("rationale"):
            lines.append("## Rationale")
            lines.append(str(digest["rationale"]))
            lines.append("")
        if digest.get("approaches_tried"):
            lines.append("## What was tried")
            for a in digest["approaches_tried"]:
                approach = a.get("approach", "")
                outcome = a.get("outcome", "")
                reason = a.get("reason", "")
                lines.append(f"- {approach} → {outcome}: {reason}")
            lines.append("")
        if digest.get("next_steps"):
            lines.append("## Next steps")
            for i, step in enumerate(digest["next_steps"], 1):
                lines.append(f"{i}. {step}")
            lines.append("")
        if digest.get("key_decisions"):
            lines.append("## Key decisions")
            for d in digest["key_decisions"]:
                lines.append(f"- {d.get('decision', '')}: {d.get('rationale', '')}")
            lines.append("")
        if digest.get("files_touched"):
            lines.append("## Files touched")
            for f in digest["files_touched"]:
                summary = f" — {f['summary']}" if f.get("summary") else ""
                lines.append(f"- {f.get('path', '')} ({f.get('action', '')}){summary}")
            lines.append("")
        if digest.get("open_questions"):
            lines.append("## Open questions")
            for q in digest["open_questions"]:
                lines.append(f"- {q}")
            lines.append("")
        if digest.get("source_refs"):
            lines.append("## Source references")
            for ref in digest["source_refs"]:
                lines.append(f"- {ref}")
            lines.append("")
    else:
        lines.append("No continuation note present.")
        lines.append("")
    # Availability
    availability = payload.get("availability")
    if availability:
        lines.append("## Environment")
        ws = availability.get("workspace_available")
        lines.append(f"Workspace: {'available' if ws else 'missing'}")
        for a in availability.get("artifacts", []):
            status = "✓" if a.get("available") else "✗"
            lines.append(f"- {status} {a.get('ref', '')}")
        lines.append("")
    # Next step
    next_step = payload.get("next_step")
    if next_step:
        lines.append(f"## Next step\n{next_step}")
    return "\n".join(lines)


def register_todo_continuation(subparsers, add_format):
    parser = subparsers.add_parser(
        "handoff", help="Explicit cross-agent Todo handoff: prepare, inspect, adopt (local lease-free authority)."
    )
    # Note: we don't use add_format here because we need a custom --format
    # with a 'digest' choice. We add it manually below.
    parser.add_argument("action", choices=["prepare", "inspect", "adopt"])
    parser.add_argument("--goal-id", required=True)
    parser.add_argument("--todo-id", required=True)
    parser.add_argument("--agent-id", required=True)
    parser.add_argument("--session-id", required=True, help="Current host session identifier; provenance, not authorization.")
    parser.add_argument("--operation-id", help="Stable retry identity, required for prepare/adopt.")
    parser.add_argument("--expected-revision", help="Exact revision from inspect, required for prepare/adopt.")
    parser.add_argument("--rationale", help="Decision rationale (legacy, prepare only). Prefer --from-context for rich handoff.")
    parser.add_argument("--source-ref", action="append", default=[], help="Compact source pointer (legacy, prepare only). Prefer --from-context.")
    parser.add_argument("--workspace", default=".", help="Target workspace to check locally; not persisted.")
    parser.add_argument("--artifact", action="append", default=[], help="Required workspace-relative artifact to check locally.")
    parser.add_argument("--target-agent-id", help="For adopt: the registered agent to hand off to (default: current agent).")
    parser.add_argument("--from-context", help="Path to a JSON file containing rich handoff context (work_summary, approaches_tried, next_steps, files_touched, key_decisions, open_questions).")
    parser.add_argument("--format", dest="handoff_format", choices=["markdown", "json", "digest"],
        help="Output format. 'digest' renders inspect as a readable handoff summary (inspect only).")


def handle_todo_continuation(args, *, registry_path, runtime_root_arg, output_format, print_payload):
    if args.command != "handoff":
        return None
    try:
        if args.action != "inspect" and (not args.operation_id or not args.expected_revision):
            raise ValueError("prepare/adopt require --operation-id and --expected-revision; reuse both on retry")
        if args.action != "prepare" and (args.rationale or args.source_ref):
            raise ValueError("decision rationale and source references belong to prepare")
        if args.action != "adopt" and args.target_agent_id:
            raise ValueError("--target-agent-id is only valid for adopt action")
        if args.action != "inspect" and getattr(args, "handoff_format", None) == "digest":
            raise ValueError("--format digest is only valid for inspect action")
        registry = load_registry(registry_path)
        root = resolve_runtime_root(registry, runtime_root_arg)
        # Build the payload. If --from-context is provided, read the JSON file
        # and merge its fields into the request.
        request = {
            "runtime_root": str(root),
            "goal_id": args.goal_id,
            "todo_id": args.todo_id,
            "agent_id": args.agent_id,
            "registered_agents": registered_agent_ids_from_registry(registry_path, args.goal_id),
            "action": args.action,
            "session_id": args.session_id,
            "operation_id": args.operation_id,
            "expected_provider_revision": args.expected_revision,
            "workspace": str(Path(args.workspace).expanduser().resolve()),
            "artifacts": args.artifact,
            "target_agent_id": args.target_agent_id,
        }
        if args.action == "prepare":
            if getattr(args, "from_context", None):
                context_path = Path(args.from_context).expanduser().resolve()
                with open(context_path, "r", encoding="utf-8") as f:
                    context = json.load(f)
                # Validate context is an object with only context-only keys
                # BEFORE placing it in the request. This prevents a context
                # file from overriding CLI-derived command/identity/authority
                # metadata (action, agent_id, goal_id, etc.).
                request["context"] = _validate_context_payload(context)
            else:
                # Legacy: pass through rationale and source_refs.
                if args.rationale:
                    request["rationale"] = args.rationale
                if args.source_ref:
                    request["source_refs"] = args.source_ref
        payload = effect_runtime_result("coordination.local_authority.todo_continuation", request)
    except (ValueError, RuntimeError, OSError, json.JSONDecodeError) as exc:
        payload = {"ok": False, "status": "failed",
            "reason_code": "invalid_continuation_request", "reason": str(exc)}
    # Determine output format.
    handoff_format = getattr(args, "handoff_format", None)
    if handoff_format:
        fmt = handoff_format
    else:
        fmt = output_format(args)
    if fmt == "digest":
        print(_render_digest(payload))
    else:
        print_payload(payload, fmt, lambda value: json.dumps(value, indent=2, ensure_ascii=False))
    return 0 if payload.get("ok") else 1
